---
title: "Evolutionary Genomics & Bioinformatics"
author: "Marisa Abundis"
output:
  prettydoc::html_pretty:
    theme: lenoids
    highlight: github
bibliography: data/references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<center>
![](images/bunnyPortraitIcon.png)
</center>

## **LAB FILES**

* Lab 1 : [Intro to R](labs/lab1_marisaabundis.html)
* Lab 2 : [Git, Github, and Rstudio Projects](labs/lab2_MarisaAbundis.html)
* Lab 3A : [Starting with the Data](lab3A.html)
* Lab 3B : [Manipulating Data](lab3B.html)
* Lab 4A : [ggplot2 Examples and Challenges](lab4A.html)
* Lab 4B : [Data Wrangling and Graphing COVID-19 Reporting Data](lab4B.html)
* Lab 5A : [Data Wrangling and Graphing COVID-19 Reporting Data (pt.2)](lab5.html)
* Lab 5B : [Interactive Graphs and Animations](lab5B.html)
* Lab 6 : [Data Maps and Interactive Graphs from the COVID-19 Reporting Data](lab6.html)
* Lab 7 : [RNA-Seq workflow: Gene-Level Exploratory Analysis & Differential Expression](Lab7.html)
* Lab 8 : [Sequence Alignment, Phylogenetic Analysis & Tree Visualization](Lab8.html)
* Lab 9 : [Programming in R - Regular Expressions](lab9.html)
* Lab 9x : [Programming in R - Control Structures & Functions](lab9x.html)
* Lab 10 : [Population genomics based on high throughput sequencing](lab10.html)
* Lab 11: [Microbiome analysis using dada2 and phyloseq](lab11.html)
* Lab 12: [Network analysis using Cytoscape & RCy3](lab12.html)
* Lab 13: [Sprucing up your Rmd & GitHub Page](lab13.html)

*Application written in R [@RCoreTeam] using the Shiny framework [@Chang2015].*


## **REFERENCES**
